### DISCUSSION TITLE
post/fire-and-forget-audio-playback-with

### DISCUSSION THREAD

**Discussion by giscus** [markheath/soundcode-blog-comments#46] [Announcements]
# post/fire-and-forget-audio-playback-with

Mark Heath's Development Blog

https://www.markheath.net/post/fire-and-forget-audio-playback-with

**Comment by nycos62** [comment]
Hello, thank you for the code, I know its a bit late, but I wish I could add a sort of channel per CachedSample: If you play again the same cachedsample , it stop the previous cachedsample playing and start the sample again.  also an event when cached sample playing stop. Do you have an idea on how to proceed ? ( for the stop playing notification I know I can estimate cachedsample length ) I'm trying to make a DJ UI like Pioneer DJS-1000

**Comment by markheath** [comment]
has obviously been a while since I wrote that article, but I can't see why it wouldn't be able to play multiple instanancs of the same cached sample simultaneously as it creates a new `CachedSoundSampleProvider` for each one

**Reply by nycos62** [reply]
thank you for the advise, just what I needed :

*ID added to CachedSamples*
```
class CachedSound
    {
        public string Id { get; private set; }
        public float[] AudioData { get; private set; }
        public WaveFormat WaveFormat { get; private set; }
        public CachedSound(string audioFileName)
        {
            using (var audioFileReader = new AudioFileReader(audioFileName))
            {
                // TODO: could add resampling in here if required
                WaveFormat = audioFileReader.WaveFormat;
                var wholeFile = new List<float>((int)(audioFileReader.Length / 4));
                var readBuffer = new float[audioFileReader.WaveFormat.SampleRate * audioFileReader.WaveFormat.Channels];
                int samplesRead;
                while ((samplesRead = audioFileReader.Read(readBuffer, 0, readBuffer.Length)) > 0)
                {
                    wholeFile.AddRange(readBuffer.Take(samplesRead));
                }
                AudioData = wholeFile.ToArray();
                Id = Guid.NewGuid().ToString();
            }
        }
    }
```

*Position made public*
```
class CachedSoundSampleProvider : ISampleProvider
    {
        private readonly CachedSound cachedSound;
        public long position;

        public CachedSoundSampleProvider(CachedSound cachedSound)
        {
            this.cachedSound = cachedSound;
        }

        public int Read(float[] buffer, int offset, int count)
        {
            Console.WriteLine("offset:" + offset + " position:" + position);
            var availableSamples = cachedSound.AudioData.Length - position;
            var samplesToCopy = Math.Min(availableSamples, count);
            Array.Copy(cachedSound.AudioData, position, buffer, offset, samplesToCopy);
            position += samplesToCopy;
            return (int)samplesToCopy;
        }

        public WaveFormat WaveFormat { get { return cachedSound.WaveFormat; } }
    }
```


*Monophonic attribute added* ```AudioPlaybackEngine.Instance.Monophonic = true;```
*then the same instance of a cachedSample is no longer played multiple times simultaneously*


```
class AudioPlaybackEngine : IDisposable
    {
        private readonly IWavePlayer outputDevice;
        private readonly MixingSampleProvider mixer;
        public bool Monophonic = false;
        private System.Collections.Hashtable HT_providers = new System.Collections.Hashtable();

        public AudioPlaybackEngine(int sampleRate = 44100, int channelCount = 2)
        {
            outputDevice = new WaveOutEvent();
            mixer = new MixingSampleProvider(WaveFormat.CreateIeeeFloatWaveFormat(sampleRate, channelCount));
            mixer.ReadFully = true;            
            outputDevice.Init(mixer);
            outputDevice.Play();
        }

        public void PlaySound(string fileName)
        {
            var input = new AudioFileReader(fileName);
            AddMixerInput(new AutoDisposeFileReader(input));
        }

        private ISampleProvider ConvertToRightChannelCount(ISampleProvider input)
        {
            if (input.WaveFormat.Channels == mixer.WaveFormat.Channels)
            {
                return input;
            }
            if (input.WaveFormat.Channels == 1 && mixer.WaveFormat.Channels == 2)
            {
                return new MonoToStereoSampleProvider(input);
            }
            throw new NotImplementedException("Not yet implemented this channel count conversion");
        }

        public void PlaySound(CachedSound sound)
        {
            if (!Monophonic)
                AddMixerInput(new CachedSoundSampleProvider(sound));
            else
            {
                CachedSoundSampleProvider cssp = GetProvider(sound);
                mixer.RemoveMixerInput(cssp);
                cssp.position = 0;                
                AddMixerInput(cssp);
            }
        }

        private void AddMixerInput(ISampleProvider input)
        {
            mixer.AddMixerInput(ConvertToRightChannelCount(input));
        }

        private CachedSoundSampleProvider GetProvider(CachedSound sound)
        {
            if (HT_providers.ContainsKey(sound.Id))
            {
                return (CachedSoundSampleProvider)HT_providers[sound.Id];
            }
            else
            {
                CachedSoundSampleProvider cssp = new CachedSoundSampleProvider(sound);
                HT_providers.Add(sound.Id, cssp);
                return cssp;
            }
        }

        public void Dispose()
        {
            outputDevice.Dispose();
        }

        public static readonly AudioPlaybackEngine Instance = new AudioPlaybackEngine(44100, 2);
    }
```

**Comment by ITheP** [comment]
Reminder as well (as I forgot) - if playing lots of sounds very quickly (I needed this for something that sounds like a Geiger counter/spinning wheel that clicks with each spoke going past) - don't forget to set a suitable DesiredLatency/NumberOfBuffers when you create the WaveOutEvent, else you might find thing's don't quite work out and multiple new sounds might start being played at the same time rather than at separate times. (And reminder if the latency is too low, you can get into audio problems- I'm using 50ms and 4 buffers but that might not work for you).

**Reply by olegtutaev** [reply]
I don't understand, how to set 50ms and 4 buffers? I will be glad to see your answer

**Comment by kimjwatt5213** [comment]
I'm using the Fire and Forget, and it works great for 44100 sample rate songs. With a 48000 sample rate song it is crashing on the line of code in AddMixerInput() "All mixer inputs must have the same WaveFormat". In the AudioPlaybackEngine constructor I'm passing in the 48000 sampleRate and the mixer is set correctly, but in the method PlaySound(), the AudioReader.WaveFormat.SampleRate is 44100. It is readonly so I cannot change it.

**Reply by kimjwatt5213** [reply]
Can you add a sampleRate parameter in the AudioReader constructor, or a settable property?

**Reply by markheath** [reply]
Yes, there is a constructor parameter where you can set the sample rate of the files you'll be playing. If you want to play a mix of sample rates then you either need to do sample rate conversion yourself, or just open multiple WaveOut devices, one for each sample rate you need

**Reply by kimjwatt5213** [reply]
Thanks Mark. I'm not seeing that parameter. The AudioFileReader has one constructor that only takes a filename. The AudioPlaybackEngine takes the sample rate, but there is no way to get that into the AudioFileReader.

**Reply by markheath** [reply]
Set the AudioPlaybackEngine to the sample rate of the file you are going to play. If you need to play files with different sample rates, you need more than one AudioPlaybackEngine (or to resample audio)

**Reply by kimjwatt5213** [reply]
I'm disposing the engine and creating a new one each song with the sample rate and channels that I get from the metadata. Here's a song you can try that is 48000. Create the AudioPlaybackEngine with a sample rate of 48000 and 2 channels. It will throw an exception in the method PlaySound(). https://kimwatt.com/files/song.zip

**Reply by markheath** [reply]
In that case, you don't need the AudioPlaybackEngine as you're playing songs, not mixing multiple short sounds which is what that engine is intended for, so I'd ditch that and pass an AudioFileReader directly into WaveOutEvent or WasapiOut instead

**Reply by kimjwatt5213** [reply]
Thanks again Mark - that sure made it simple and works great! Thanks for taking the time to respond to me.
